x-pack/legacy/plugins/apm/server/lib/service_map/get_service_map.ts

```diff
@@ -0,0 +1,164 @@
+
+export interface IEnvOptions {
+  setup: Setup & SetupTimeRange & SetupUIFilters;
+  serviceName?: string;
+  environment?: string;
+}
+
+export type ServiceMapAPIResponse = PromiseReturnType<typeof getServiceMap>;
+export async function getServiceMap({
+  setup,
+  serviceName,
+  environment
+}: IEnvOptions) {
+  const { start, end, client, config, uiFiltersES } = setup;
+
+  const params = {
+    index: config['xpack.apm.serviceMapIndexPattern'],
+    body: {
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            { range: rangeFilter(start, end) },
+            { exists: { field: 'connection.type' } },
+            ...uiFiltersES
+          ]
+        }
+      },
+      aggs: {
+        conns: {
+          composite: {
+            sources: [
+              { 'service.name': { terms: { field: 'service.name' } } },
+              {
+                'service.environment': {
+                  terms: { field: 'service.environment', missing_bucket: true }
+                }
+              },
+              {
+                'destination.address': {
+                  terms: {
+                    field: 'destination.address'
+                  }
+                }
+              },
+              { 'connection.type': { terms: { field: 'connection.type' } } }, // will filter out regular spans etc.
+              {
+                'connection.subtype': {
+                  terms: { field: 'connection.subtype', missing_bucket: true }
+                }
+              }
+            ],
+            size: 1000
+          },
+          aggs: {
+            dests: {
+              terms: {
+                field: 'callee.name'
+              },
+              aggs: {
+                envs: {
+                  terms: {
+                    field: 'callee.environment',
+                    missing: ''
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  };
+
+  if (serviceName || environment) {
+    const upstreamServiceName = serviceName || '*';
+
+    let upstreamEnvironment = '*';
+    if (environment) {
+      upstreamEnvironment =
+        environment === ENVIRONMENT_NOT_DEFINED ? 'null' : environment;
+    }
+
+    params.body.query.bool.filter.push({
+      wildcard: {
+        ['connection.upstream.list']: `${upstreamServiceName}/${upstreamEnvironment}`
+      }
+    });
+  }
+
+  // @ts-ignore
+  const { aggregations } = await client.search(params);
+  const buckets: Array<{
+    key: {
+      'service.name': string;
+      'service.environment': string | null;
+      'destination.address': string;
+      'connection.type': string;
+      'connection.subtype': string | null;
+    };
+    dests: {
+      buckets: Array<{
+        key: string;
+        envs: {
+          buckets: Array<{ key: string }>;
+        };
+      }>;
+    };
+  }> = idx(aggregations, _ => _.conns.buckets) || [];
```

Comment:
@dgieselaar is this something our magic agg types should be able to infer instead of them being spelled out?

Reply:
Looks like this is no longer an issue, correct?